File: Lab3/zad1/cpp/main.cpp
#include "animal.hpp"
#include "myfactory.hpp"

#include <iostream>

void printGreeting(Animal& animal) {
    std::cout << animal.name() << " pozdravlja: " << animal.greet() << '\n';
}

void printMenu(Animal& animal) {
    std::cout << animal.name() << " voli " << animal.menu() << '\n';
}

int main() {
    auto& factory = MyFactory::instance();
    auto ids = factory.getIds<Animal>();
    for (const std::string& id : ids) {
        std::string pet = "Ljubimac " + id;
        auto animal = factory.create<Animal>(id, pet);
        if (animal) {
            printGreeting(*animal);
            printMenu(*animal);
        }
    }

    return 0;
}

File: Lab3/zad1/python/main.py
import os
from importlib import import_module

def myfactory(module_name):
    # Dynamically import the module and get the class from it
    return getattr(import_module('plugins.' + module_name), module_name)

def printGreeting(pet):
    print(f"{pet.name()} pozdravlja: {pet.greet()}")

def printMenu(pet):
    print(f"{pet.name()} voli {pet.menu()}.")

def main():
    pets = []
    for mymodule in os.listdir('plugins'):
        module_name, module_ext = os.path.splitext(mymodule)
        if module_ext == '.py':
            pet = myfactory(module_name)(f"Ljubimac {len(pets)}")
            pets.append(pet)

    for pet in pets:
        printGreeting(pet)
        printMenu(pet)

if __name__ == '__main__':
    main()